### PR TITLE
Add `torch.compile()` support for BAE’s PyPose `LieTensor` operations and sparse Jacobian traversal

### PR DESCRIPTION
**Pull request by ZihangFang**
## Summary

This PR adds `torch.compile()` support for BAE’s PyPose `LieTensor` operations and sparse Jacobian traversal, including indexed bundle-adjustment parameters.

## What changed

- Adds a compatibility shim for PyPose `LieTensor` that removes TorchDynamo graph breaks caused by:

  - `Tensor.as_subclass()` when converting a `LieTensor` to a base tensor
  - `hasattr()` introspection in `__torch_function__`
  - dynamic `requires_grad` tensor construction in `vec2skew`
  - incorrect compiled dispatch of Lie group `@` operations

- Makes the existing ambient-gradient monkey patch compile-compatible, including SO(3)/SE(3) `Exp`, `Log`, `Act`, `Mul`, and `Inv`.

- Replaces the legacy `id(tensor) → edge` `optrace` dictionary with a direct tuple representation. Each tracked tensor has exactly one incoming edge, so the dictionary was redundant.

- Adds compile-safe sparse Jacobian traversal through `jacobian_components()`:

  - avoids mutating tensor attributes during traversal
  - supports map, index, concatenation, and repeated parameter contributions
  - preserves LieTensor metadata needed by `jacrev`
  - keeps indexed camera and point gathers inside the compiled graph

- Adds `materialize_jacobian_components()` to construct sparse BSR tensors after the compiled call. PyTorch cannot currently return sparse BSR tensors through FakeTensor/AOTAutograd, so compiled code returns their dense component tensors instead. Single-component BSR materialization is zero-copy.

- Updates LM parameter retraction and local-Jacobian trimming to operate through base tensor aliases, avoiding mixed tensor-subclass dispatch.

- Documents activation through:

```bash
export BAE_USE_PYPOSE_TORCH_COMPILE=1
```

The existing `BAE_USE_PYPOSE_AMBIENT_GRAD=1` option also installs the compile compatibility shim.

## Indexing and fusion

Indexed camera and point arguments are stored compactly as `(parent, index)` trace metadata and reconstructed inside the compiled graph. This allows Inductor to fuse or otherwise optimize indexed loads with downstream LieTensor and Jacobian operations.

`fullgraph=True` guarantees complete graph capture, but it does not guarantee a particular kernel-fusion or intermediate-buffer strategy.

## Validation

- Fullgraph LieTensor forward and gradient tests
- Fullgraph `jacrev` for SO(3)/SE(3) operations
- Eager-versus-compiled sparse Jacobian comparisons
- Index, concatenation, repeated-branch, BAL, and PGO traversal coverage
- Compiled BAL residual and sparse Jacobian components on CPU and CUDA
- Zero-copy BSR value materialization verified by matching storage pointers
- 22 graph/PyPose compile tests passed
- 2 compiled BAL indexing tests passed on CPU and CUDA

In the tested BA workload, residual and Jacobian-component compilation produced one graph with no graph breaks and reduced runtime from approximately 10.35 ms to 2.21 ms, about a 4.7× speedup.

### CODE REVIEW

File: bae/autograd/graph.py [hunk 1]
```diff
@@ -178,15 +210,251 @@ def _vmap_in_dims(args):
...
+        if output.ndim == 1:
+            values = torch.ones(
+                (output.shape[0], 1, 1),
+                device=output.device,
+                dtype=output.dtype,
+            )
+        else:
+            block_dim = output.shape[-1]
+            eye = torch.eye(block_dim, device=output.device, dtype=output.dtype)
+            values = eye.unsqueeze(0).repeat(output.shape[0], 1, 1)
```
[the remark this answers is not in the document]
Comment (ZihangFang) [reply bae/autograd/graph.py:261]: This is fine because we materialize the Jacobian block anyway in BSR

**Comment by zitongzhan** [comment]
@gemini-code-assist review

File: bae/autograd/graph.py [hunk 2]
```diff
@@ -178,15 +210,251 @@ def _vmap_in_dims(args):
...
+                    (output.shape[0], 1, 1),
+                    device=output.device,
+                    dtype=output.dtype,
+                )
+            else:
+                block_dim = output.shape[-1]
+                eye = torch.eye(
+                    block_dim, device=output.device, dtype=output.dtype
+                )
+                values = eye.unsqueeze(0).repeat(output.shape[0], 1, 1)
```
[the remark this answers is not in the document]
Comment (zitongzhan) [reply bae/autograd/graph.py:360]: Not planned

File: bae/autograd/function.py [hunk 3]
```diff
@@ -69,23 +83,16 @@ def _find_tracking_source(values, cls):
 def _unwrap_tracking_tensor(value):
     if not isinstance(value, TrackingTensor):
         return value
+    base = torch.Tensor(value)
```
[the remark this answers is not in the document]
Comment (zitongzhan) [reply bae/autograd/function.py:86]: This is to make sure torch dynamo can compile
